lmdeploy/pytorch/models/minicpm3.py

```diff
@@ -0,0 +1,632 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import math
+from typing import Any, Iterable, List, Optional, Tuple
+
+import torch
+import torch.distributed as dist
+from torch import nn
+from transformers.configuration_utils import PretrainedConfig
+
+from lmdeploy.pytorch.model_inputs import StepContext, StepContextManager
+from lmdeploy.pytorch.nn import (Attention, RMSNorm, RopeType, SiluAndMul,
+                                 build_rotary_embedding)
+from lmdeploy.pytorch.nn.linear import (build_colwise_linear,
+                                        build_merged_colwise_linear,
+                                        build_rowwise_linear)
+from lmdeploy.pytorch.weight_loader.model_weight_loader import load_weight
+
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., :x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2:]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids, unsqueeze_dim=1):
```

Comment:
We have provide apply rotary and rotary embedding op in `pytorch.nn`, Which is fused and would provide better performance.